supervision/detection/utils.py

```diff
@@ -691,23 +691,51 @@ def merge_data(
     if not data_list:
         return {}
 
-    all_keys_sets = [set(data.keys()) for data in data_list]
-    if not all(keys_set == all_keys_sets[0] for keys_set in all_keys_sets):
-        raise ValueError("All data dictionaries must have the same keys to merge.")
-
     for data in data_list:
-        lengths = [len(value) for value in data.values()]
-        if len(set(lengths)) > 1:
+        lengths_set = [len(value) for value in data.values()]
+        if len(set(lengths_set)) > 1:
             raise ValueError(
                 "All data values within a single object must have equal length."
             )
 
-    merged_data = {key: [] for key in all_keys_sets[0]}
-
+    all_keys: Set[str] = set()
     for data in data_list:
-        for key in merged_data:
-            merged_data[key].append(data[key])
+        all_keys.update(data.keys())
+
+    # Naively merging entries and then validating length comes with a problem:
```

Comment:
Please remove the comment. Comments are allowed only when we are doing something non-standard, such as a hack that is very non-obvious. Writing comments that describe the code is a bad practice.

Use variable and function names, as well as unit tests, to make the code more self-descriptive.